### PR TITLE
util/attr, pingpong: Add support of FI_MSG_PREFIX mode

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
- The 1st pathc adds support of `FI_MSG_PREFIX` mode in pingpong test.
- In order to support this mode for providers that use util/attr code, the 2nd patch adds support of `FI_MSG_PREFIX` mode

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@bturrubiates @jsquyres Can I ask you verify the changes to pingpong test for the usnic/DGRAM? Thanks in advance!

File: prov/util/src/util_attr.c [hunk 1]
```diff
@@ -622,7 +622,7 @@ int ofi_check_ep_attr(const struct util_prov *util_prov, uint32_t api_version,
 		      const struct fi_info *user_info)
 {
 	const struct fi_ep_attr *prov_attr = prov_info->ep_attr;
-	const struct fi_ep_attr *user_attr = user_info->ep_attr;
+	struct fi_ep_attr *user_attr = user_info->ep_attr;
```
Comment (shefty) [inline prov/util/src/util_attr.c]: This is a 'check' call, we shouldn't modify the user's attributes.  That should be done as part of an 'alter' call.

Comment (gladkovdmitry17) [reply prov/util/src/util_attr.c]: Thanks! Done.

File: prov/util/src/util_attr.c [hunk 2]
```diff
@@ -1078,8 +1078,8 @@ static void fi_alter_domain_attr(struct fi_domain_attr *attr,
 }
 
 static void fi_alter_ep_attr(struct fi_ep_attr *attr,
-			     const struct fi_ep_attr *hints,
-			     uint64_t info_caps)
+			     struct fi_ep_attr *hints,
```
Comment (shefty) [inline prov/util/src/util_attr.c]: hints must remain const.  We should not be altering the user's provided hints.  We are only setting 'attr' in this function.

File: prov/util/src/util_attr.c [hunk 3]
```diff
@@ -1096,6 +1096,8 @@ static void fi_alter_ep_attr(struct fi_ep_attr *attr,
 		attr->tx_ctx_cnt = hints->tx_ctx_cnt;
 	if (hints->rx_ctx_cnt)
 		attr->rx_ctx_cnt = hints->rx_ctx_cnt;
+	if (user_modes & FI_MSG_PREFIX)
+		hints->msg_prefix_size = attr->msg_prefix_size;
```
Comment (shefty) [inline prov/util/src/util_attr.c]: I *think* this is the flow before we reach this spot:
1. If the provider requires FI_MSG_PREFIX, the hints must also.  That check should have already been done.
1a. Duping the provider's info structure will have set msg_prefix_size
2. If the provider _may_ use FI_MSG_PREFIX, hints->mode is not verified for this bit.
2a. Duping the provider's info will have msg_prefix_size set.
2b. We need to clear msg_prefix_size here if hints->mode does NOT have FI_MSG_PREFIX

That is, I believe a provider that requires FI_MSG_PREFIX should work with the existing code.  For a provider that _wants_ but does not require FI_MSG_PREFIX, it should NOT set FI_MSG_PREFIX as a mode bit, but should set msg_prefix_size.  In this case, I think we need something like this:
```
if (!(user_mode &FI_MSG_PREFIX))
    attr->msg_prefix_size = 0;
```
We need to check the entire source flow to confirm that this is what would happen.

**Comment by shefty** [comment]
Changes to pingpong test updated to latest master moved to PR #3882.

**Pull request closed without merging**